inject/src/main/java/io/micronaut/context/env/DefaultConfigurationPath.java

```diff
@@ -0,0 +1,481 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context.env;
+
+import io.micronaut.context.annotation.ConfigurationReader;
+import io.micronaut.context.annotation.EachProperty;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.core.value.PropertyResolver;
+import io.micronaut.inject.BeanDefinition;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Implementation of {@link ConfigurationPath}.
+ *
+ * @since 4.0.0
+ */
+final class DefaultConfigurationPath implements ConfigurationPath {
+    private final LinkedList<ConfigurationSegment> list = new LinkedList<>();
```

Comment:
It might be better to use ArrayDeque, which should be more performant. We might want to also use it for `AbstractBeanResolutionContext.DefaultPath`

Reply:
There is no `subList` method in ArrayDeque